api/types/resource.go

```diff
@@ -704,3 +704,51 @@ func GetKind(v any) string {
 
 	return ""
 }
+
+// GetRevision returns the revision if one can be obtained.
```

Comment:
Better yet:

```suggestion
// GetRevision returns the revision if one can be obtained.
// Works with both [Resource] and [ResourceMetadata] instances, returns empty
// otherwise.
```

For the other methods we should tweak the last sentence accordingly. Do you think it's worth panicking if we don't match a known interface, as to avoid difficult to debug bugs?